ouroboros-network/src/Ouroboros/Network/Server/ConnectionTable.hs

```diff
@@ -0,0 +1,214 @@
+
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.Server.ConnectionTable (
+      ConnectionTable
+    , ConnectionTableRef (..)
+    , ValencyCounter
+
+    , newConnectionTable
+    , refConnection
+    , addConnection
+    , removeConnection
+    , newValencyCounter
+    , addValencyCounter
+    , remValencyCounter
+    , waitValencyCounter
+    , readValencyCounter
+    ) where
+
+import           Control.Monad (when)
+import           Control.Monad.Class.MonadSTM
+--import           Control.Tracer XXX Not Yet
+import qualified Data.Map.Strict as M
+import           Data.Set (Set)
+import qualified Data.Set as S
+import qualified Network.Socket as Socket
+import           Text.Printf
+
+-- A ConnectionTable represent a set of connections that is shared between
+-- servers and subscription workers. Its main purpose is to avoid the creation of duplicate
+-- connections (especially connections with identical source address, destination address, source
+-- port and destination port which would be rejected by the kernel anyway.).
+-- It is only used for bookkeeping, the sockets represented by the connections are not accessable
+-- through this structure.
```

Comment:
Non essential, but @dcoutts pointed to me that putting a single empty `--` comment at the end makes the code easier to read (it stands out more from the documentation).

Reply:
Will add `--´ line.